docs/CONFIGURATION.md
# Configuration

Thunderdome may be configured through environment variables or via a yaml file `config.yaml`
located in one of:

* `/etc/thunderdome/`
* `$HOME/.config/thunderdome/`
* Current working directory

### Example yaml configuration file

```
http:
  domain: thunderdome.dev
db:
  host: localhost
  port: 5432
  user: thor
  pass: odinson
  name: thunderdome
```

## Required configuration items

For Thunderdome to work correctly the following configuration items are required:

| Option                | Environment Variable | Description                                                                              | Default Value     |
|-----------------------|----------------------|------------------------------------------------------------------------------------------|-------------------|
| `http.domain`         | APP_DOMAIN           | The domain/base URL for this instance of Thunderdome. Used for creating URLs in emails. | thunderdome.dev   |
| `http.cookie_hashkey` | COOKIE_HASHKEY       | Secret used to make secure cookies secure.                                               | strongest-avenger |
| `config.aes_hashkey`  | CONFIG_AES_HASHKEY   | Secret used to encrypt passcode fields (e.g. Battle JoinCode, LeaderCode).               | therevengers      |

### Database configuration

Thunderdome uses a Postgres database to store all data, the following configuration options exist:

| Option                     | Environment Variable | Description                                | Default Value           |
| -------------------------- | -------------------- | ------------------------------------------ | ------------------------|
| `db.host`                  | DB_HOST              | Database host name.                        | db |
| `db.port`                  | DB_PORT              | Database port number.                      | 5432 |
| `db.user`                  | DB_USER              | Database user id.                          | thor |
| `db.pass`                  | DB_PASS              | Database user password.                    | odinson |
| `db.name`                  | DB_NAME              | Database instance name.                    | thunderdome |
| `db.sslmode`               | DB_SSLMODE           | Database SSL Mode (disable, allow, prefer, require, verify-ca, verify-full). | disable |

### SMTP (Mail) server configuration

Thunderdome sends emails for user registration related activities, the following configuration options exist:

| Option                     | Environment Variable | Description                                | Default Value           |
| -------------------------- | -------------------- | ------------------------------------------ | ------------------------|
| `smtp.host`                | SMTP_HOST            | Smtp server hostname.                      | localhost |
| `smtp.port`                | SMTP_PORT            | Smtp server port number.                   | 25 |
| `smtp.user`                | SMTP_USER            | Smtp server user.                          | |
| `smtp.pass`                | SMTP_PASS            | Smtp server password.                      | |
| `smtp.secure`              | SMTP_SECURE          | Set to authenticate with the Smtp server.  | true |
| `smtp.identity`            | SMTP_IDENTITY        | Smtp server authorization identity. Usually unset. | |
| `smtp.sender`              | SMTP_SENDER          | From address in emails sent by Thunderdome. | no-reply@thunderdome.dev |

## Configure Admin Email

To grant Admin access to Thunderdome for the first Admin user create an account first, then set the `ADMIN_EMAIL`
environment variable or `admin.email` config file value. Upon the app starting up Thunderdome will read this value and
set that user as ADMIN role.

## Optional configuration items

| Option                                | Environment Variable                | Description                                                                                                          | Default Value                          |
|---------------------------------------|-------------------------------------|----------------------------------------------------------------------------------------------------------------------|----------------------------------------|
| `http.port`                           | PORT                                | Which port to listen for HTTP connections.                                                                           | 8080                                   |
| `http.path_prefix`                    | PATH_PREFIX                         | Prefix added to all application urls for shared domain use, in format of `/{prefix}` e.g. `/thunderdome`             |                                        |
| `http.secure_cookie`                  | COOKIE_SECURE                       | Use secure cookies or not.                                                                                           | true                                   |
| `http.backend_cookie_name`            | BACKEND_COOKIE_NAME                 | The name of the backend cookie utilized for actual auth/validation                                                   | warriorId                              |
| `http.frontend_cookie_name`           | FRONTEND_COOKIE_NAME                | The name of the cookie utilized by the UI (purely for convenience not auth)                                          | warrior                                |
| `analytics.enabled`                   | ANALYTICS_ENABLED                   | Enable/disable google analytics.                                                                                     | true                                   |
| `analytics.id`                        | ANALYTICS_ID                        | Google analytics identifier.                                                                                         | UA-140245309-1                         |
| `config.allowedPointValues`           | CONFIG_POINTS_ALLOWED               | List of available point values for creating battles.                                                                 | 0, 1/2, 2, 3, 5, 8, 13, 20, 40, 100, ? |
| `config.defaultPointValues`           | CONFIG_POINTS_DEFAULT               | List of default selected points for new battles.                                                                     | 1, 2, 3, 5, 8 , 13, ?                  |
| `config.show_warrior_rank`            | CONFIG_SHOW_RANK                    | Set to enable an icon showing the rank of a warrior during battle.                                                   | false                                  |
| `config.avatar_service`               | CONFIG_AVATAR_SERVICE               | Avatar service used, possible values see next paragraph                                                              | gravatar                               |
| `config.toast_timeout`                | CONFIG_TOAST_TIMEOUT                | Number of milliseconds before notifications are hidden.                                                              | 1000                                   |
| `config.allow_guests`                 | CONFIG_ALLOW_GUESTS                 | Whether or not to allow guest (anonymous) users.                                                                     | true                                   |
| `config.allow_registration`           | CONFIG_ALLOW_REGISTRATION           | Whether or not to allow user registration (outside Admin).                                                           | true                                   |
| `config.allow_jira_import`            | CONFIG_ALLOW_JIRA_IMPORT            | Whether or not to allow import plans from JIRA XML.                                                                  | true                                   |
| `config.default_locale`               | CONFIG_DEFAULT_LOCALE               | The default locale (language) for the UI                                                                             | en                                     |
| `config.friendly_ui_verbs`            | CONFIG_FRIENDLY_UI_VERBS            | Whether or not to use more friendly UI verbs like Users instead of Warrior, e.g. Corporate friendly                  | false                                  |
| `config.allow_external_api`           | CONFIG_ALLOW_EXTERNAL_API           | Whether or not to allow External API access                                                                          | false                                  |
| `config.user_apikey_limit`            | CONFIG_USER_APIKEY_LIMIT            | Limit users number of API keys                                                                                       | 5                                      |
| `config.show_active_countries`        | CONFIG_SHOW_ACTIVE_COUNTRIES        | Whether or not to show active countries on landing page                                                              | false                                  |
| `config.cleanup_battles_days_old`     | CONFIG_CLEANUP_BATTLES_DAYS_OLD     | How many days back to clean up old battles, e.g. battles older than 180 days. Triggered manually by Admins .         | 180                                    |
| `config.cleanup_retros_days_old`      | CONFIG_CLEANUP_RETROS_DAYS_OLD      | How many days back to clean up old retros, e.g. retros older than 180 days. Triggered manually by Admins .           | 180                                    |
| `config.cleanup_storyboards_days_old` | CONFIG_CLEANUP_STORYBOARDS_DAYS_OLD | How many days back to clean up old storyboards, e.g. storyboards older than 180 days. Triggered manually by Admins . | 180                                    |
| `config.cleanup_guests_days_old`      | CONFIG_CLEANUP_GUESTS_DAYS_OLD      | How many days back to clean up old guests, e.g. guests older than 180 days. Triggered manually by Admins.            | 180                                    |
| `config.organizations_enabled`        | CONFIG_ORGANIZATIONS_ENABLED        | Whether or not creating organizations (with departments) are enabled                                                 | true                                    |
| `auth.method`                         | AUTH_METHOD                         | Choose `normal` or `ldap` as authentication method. See separate section on LDAP configuration.                      | normal                                 |
| `feature.poker`                       | FEATURE_POKER                       | Enable or Disable Agile Story Pointing (Poker) feature                                                               | true                                   |
| `feature.retro`                       | FEATURE_RETRO                       | Enable or Disable Agile Retrospectives feature                                                                       | true                                   |
| `feature.storyboard`                  | FEATURE_STORYBOARD                  | Enable or Disable Agile Storyboard feature                                                                           | true                                   |

### Avatar Service configuration

Use the name from table below to configure a service - if not set, `goadorable` is used. Each service provides further
options which then can be configured by a warrior on the profile page. Once a service is configured, drop downs with the
different sprites become available. The table shows all supported services and their sprites. In all cases the same
ID (`ead26688-5148-4f3c-a35d-1b0117b4f2a9`) has been used creating the avatars.

| Name |           |           |           |           |           |           |           |           |           |
| ---------- | --------- | --------- | --------- | --------- | --------- | --------- | --------- | --------- | --------- |
| `goadorable` (internal)  |           |           |           |           |           |           |           |           |           |
|            | ![image](https://user-images.githubusercontent.com/846933/96212071-e4283d80-0f43-11eb-9f82-ff6c105f8b0a.png) |
| `govatar` (internal) | male | female |  |  |
|            | ![image](https://user-images.githubusercontent.com/846933/96212029-ce1a7d00-0f43-11eb-9e53-8ca13ba9d4b1.png) | ![image](https://user-images.githubusercontent.com/846933/96212031-ceb31380-0f43-11eb-832b-b02c275317a5.png) |  |  |
| `dicebear` | male | female | human | identicon | bottts | avataaars | jdenticon | gridy | code |
|            | ![image](https://avatars.dicebear.com/api/male/ead26688-5148-4f3c-a35d-1b0117b4f2a9.svg?w=48) | ![image](https://avatars.dicebear.com/api/female/ead26688-5148-4f3c-a35d-1b0117b4f2a9.svg?w=48) | ![image](https://avatars.dicebear.com/api/human/ead26688-5148-4f3c-a35d-1b0117b4f2a9.svg?w=48) | ![image](https://avatars.dicebear.com/api/identicon/ead26688-5148-4f3c-a35d-1b0117b4f2a9.svg?w=48) | ![image](https://avatars.dicebear.com/api/bottts/ead26688-5148-4f3c-a35d-1b0117b4f2a9.svg?w=48) | ![image](https://avatars.dicebear.com/api/avataaars/ead26688-5148-4f3c-a35d-1b0117b4f2a9.svg?w=48) | ![image](https://avatars.dicebear.com/api/jdenticon/ead26688-5148-4f3c-a35d-1b0117b4f2a9.svg?w=48) | ![image](https://avatars.dicebear.com/api/gridy/ead26688-5148-4f3c-a35d-1b0117b4f2a9.svg?w=48) | ![image](https://avatars.dicebear.com/api/code/ead26688-5148-4f3c-a35d-1b0117b4f2a9.svg?w=48) |
| `gravatar` | mp | identicon | monsterid | wavatar | retro | robohash | | | |
|            | ![image](https://gravatar.com/avatar/ead26688-5148-4f3c-a35d-1b0117b4f2a9?s=48&d=mp&r=g) | ![image](https://gravatar.com/avatar/ead26688-5148-4f3c-a35d-1b0117b4f2a9?s=48&d=identicon&r=g) | ![image](https://gravatar.com/avatar/ead26688-5148-4f3c-a35d-1b0117b4f2a9?s=48&d=monsterid&r=g) | ![image](https://gravatar.com/avatar/ead26688-5148-4f3c-a35d-1b0117b4f2a9?s=48&d=wavatar&r=g) | ![image](https://gravatar.com/avatar/ead26688-5148-4f3c-a35d-1b0117b4f2a9?s=48&d=retro&r=g) | ![image](https://gravatar.com/avatar/ead26688-5148-4f3c-a35d-1b0117b4f2a9?s=48&d=robohash&r=g) | | | |
| `robohash` | set1 | set2 | set3 | set4 |
|            | ![image](https://robohash.org/ead26688-5148-4f3c-a35d-1b0117b4f2a9.png?set=set1&size=48x48) | ![image](https://robohash.org/ead26688-5148-4f3c-a35d-1b0117b4f2a9.png?set=set2&size=48x48) | ![image](https://robohash.org/ead26688-5148-4f3c-a35d-1b0117b4f2a9.png?set=set3&size=48x48) | ![image](https://robohash.org/ead26688-5148-4f3c-a35d-1b0117b4f2a9.png?set=set4&size=48x48) |

### LDAP Configuration

If `auth.method` is set to `ldap`, then the Create Account function is disabled and authentication is done using LDAP.
If the LDAP server authenticates a new user successfully, the Thunderdome user profile is automatically generated.

The following configuration options are specific to the LDAP authentication method:

| Option                      | Environment Variable | Description                                                        |
| --------------------------- | -------------------- | ------------------------------------------------------------------ |
| `auth.ldap.url`             | AUTH_LDAP_URL        | URL to LDAP server, typically `ldap://host:port`                   |
| `auth.ldap.use_tls`         | AUTH_LDAP_USE_TLS    | Create a TLS connection after establishing the initial connection. |
| `auth.ldap.bindname`        | AUTH_LDAP_BINDNAME   | Bind name / bind DN for connecting to LDAP. Leave empty for no authentication. |
| `auth.ldap.bindpass`        | AUTH_LDAP_BINDPASS   | Password for the bind.                                             |
| `auth.ldap.basedn`          | AUTH_LDAP_BASEDN     | Base DN for the search for the user.                               |
| `auth.ldap.filter`          | AUTH_LDAP_FILTER     | Filter for searching for the user's login id. See below.          |
| `auth.ldap.mail_attr`       | AUTH_LDAP_MAIL_ATTR  | The LDAP property containing the user's emil address.              |
| `auth.ldap.cn_attr`         | AUTH_LDAP_CN_ATTR    | The LDAP property containing the user's name.                      |

The default `filter` is `(&(objectClass=posixAccount)(mail=%s))`. The filter must include a `%s` that will be replaced
by the user's login id. The `mail_attr` configuration option must point to the LDAP attribute containing the user's
email address. The default is `mail`. The `cn_attr` configuration option must point to the LDAP attribute containing the
user's full name. The default is `cn`.

On Linux, the parameters may be tested on the command line:

```
ldapsearch -H auth.ldap.url [-Z] -x [-D auth.ldap.bindname -W] -b auth.ldap.basedn 'auth.ldap.filter' dn auth.ldap.mail auth.ldap.cn
```

The `-Z` is only used if `auth.ldap.use_tls` is set, the `-D` and `-W` parameter is only used if `auth.ldap.bindname` is
set.

config.go
package main

import (
	"github.com/spf13/viper"
	"go.uber.org/zap"
)

// InitConfig initializes the application configuration
func InitConfig(logger *zap.Logger) {
	viper.SetConfigName("config")
	viper.SetConfigType("yaml")

	viper.AddConfigPath("/etc/thunderdome/")
	viper.AddConfigPath("$HOME/.config/thunderdome/")
	viper.AddConfigPath(".")

	viper.SetDefault("http.cookie_hashkey", "strongest-avenger")
	viper.SetDefault("http.port", "8080")
	viper.SetDefault("http.secure_cookie", true)
	viper.SetDefault("http.backend_cookie_name", "warriorId")
	viper.SetDefault("http.session_cookie_name", "sessionId")
	viper.SetDefault("http.frontend_cookie_name", "warrior")
	viper.SetDefault("http.domain", "thunderdome.dev")
	viper.SetDefault("http.path_prefix", "")

	viper.SetDefault("analytics.enabled", true)
	viper.SetDefault("analytics.id", "UA-140245309-1")

	viper.SetDefault("db.host", "db")
	viper.SetDefault("db.port", 5432)
	viper.SetDefault("db.user", "thor")
	viper.SetDefault("db.pass", "odinson")
	viper.SetDefault("db.name", "thunderdome")
	viper.SetDefault("db.sslmode", "disable")

	viper.SetDefault("smtp.host", "localhost")
	viper.SetDefault("smtp.port", "25")
	viper.SetDefault("smtp.secure", true)
	viper.SetDefault("smtp.sender", "no-reply@thunderdome.dev")

	viper.SetDefault("config.aes_hashkey", "therevengers")
	viper.SetDefault("config.allowedPointValues",
		[]string{"0", "1/2", "1", "2", "3", "5", "8", "13", "20", "40", "100", "?"})
	viper.SetDefault("config.defaultPointValues",
		[]string{"1", "2", "3", "5", "8", "13", "?"})
	viper.SetDefault("config.show_warrior_rank", false)
	viper.SetDefault("config.avatar_service", "gravatar")
	viper.SetDefault("config.toast_timeout", 1000)
	viper.SetDefault("config.allow_guests", true)
	viper.SetDefault("config.allow_registration", true)
	viper.SetDefault("config.allow_jira_import", true)
	viper.SetDefault("config.default_locale", "en")
	viper.SetDefault("config.friendly_ui_verbs", false)
	viper.SetDefault("config.allow_external_api", true)
	viper.SetDefault("config.user_apikey_limit", 5)
	viper.SetDefault("config.show_active_countries", false)
	viper.SetDefault("config.cleanup_battles_days_old", 180)
	viper.SetDefault("config.cleanup_guests_days_old", 180)
	viper.SetDefault("config.cleanup_retros_days_old", 180)
	viper.SetDefault("config.cleanup_storyboards_days_old", 180)
	viper.SetDefault("config.organizations_enabled", true)

	// feature flags
	viper.SetDefault("feature.poker", true)
	viper.SetDefault("feature.retro", true)
	viper.SetDefault("feature.storyboard", true)

	viper.SetDefault("auth.method", "normal")
	viper.SetDefault("auth.ldap.url", "")
	viper.SetDefault("auth.ldap.use_tls", true)
	viper.SetDefault("auth.ldap.bindname", "")
	viper.SetDefault("auth.ldap.bindpass", "")
	viper.SetDefault("auth.ldap.basedn", "")
	viper.SetDefault("auth.ldap.filter", "(&(objectClass=posixAccount)(mail=%s))")
	viper.SetDefault("auth.ldap.mail_attr", "mail")
	viper.SetDefault("auth.ldap.cn_attr", "cn")

	viper.BindEnv("http.cookie_hashkey", "COOKIE_HASHKEY")
	viper.BindEnv("http.port", "PORT")
	viper.BindEnv("http.secure_cookie", "COOKIE_SECURE")
	viper.BindEnv("http.backend_cookie_name", "SECURE_COOKIE_NAME")
	viper.BindEnv("http.session_cookie_name", "SESSION_COOKIE_NAME")
	viper.BindEnv("http.frontend_cookie_name", "FRONTEND_COOKIE_NAME")
	viper.BindEnv("http.domain", "APP_DOMAIN")
	viper.BindEnv("http.path_prefix", "PATH_PREFIX")

	viper.BindEnv("analytics.enabled", "ANALYTICS_ENABLED")
	viper.BindEnv("analytics.id", "ANALYTICS_ID")
	viper.BindEnv("admin.email", "ADMIN_EMAIL")

	viper.BindEnv("db.host", "DB_HOST")
	viper.BindEnv("db.port", "DB_PORT")
	viper.BindEnv("db.user", "DB_USER")
	viper.BindEnv("db.pass", "DB_PASS")
	viper.BindEnv("db.name", "DB_NAME")
	viper.BindEnv("db.sslmode", "DB_SSLMODE")

	viper.BindEnv("smtp.host", "SMTP_HOST")
	viper.BindEnv("smtp.port", "SMTP_PORT")
	viper.BindEnv("smtp.secure", "SMTP_SECURE")
	viper.BindEnv("smtp.identity", "SMTP_IDENTITY")
	viper.BindEnv("smtp.user", "SMTP_USER")
	viper.BindEnv("smtp.pass", "SMTP_PASS")
	viper.BindEnv("smtp.sender", "SMTP_SENDER")

	viper.BindEnv("config.aes_hashkey", "CONFIG_AES_HASHKEY")
	viper.BindEnv("config.allowedPointValues", "CONFIG_POINTS_ALLOWED")
	viper.BindEnv("config.defaultPointValues", "CONFIG_POINTS_DEFAULT")
	viper.BindEnv("config.show_warrior_rank", "CONFIG_SHOW_RANK")
	viper.BindEnv("config.avatar_service", "CONFIG_AVATAR_SERVICE")
	viper.BindEnv("config.toast_timeout", "CONFIG_TOAST_TIMEOUT")
	viper.BindEnv("config.allow_guests", "CONFIG_ALLOW_GUESTS")
	viper.BindEnv("config.allow_registration", "CONFIG_ALLOW_REGISTRATION")
	viper.BindEnv("config.allow_jira_import", "CONFIG_ALLOW_JIRA_IMPORT")
	viper.BindEnv("config.default_locale", "CONFIG_DEFAULT_LOCALE")
	viper.BindEnv("config.friendly_ui_verbs", "CONFIG_FRIENDLY_UI_VERBS")
	viper.BindEnv("config.allow_external_api", "CONFIG_ALLOW_EXTERNAL_API")
	viper.BindEnv("config.user_apikey_limit", "CONFIG_USER_APIKEY_LIMIT")
	viper.BindEnv("config.show_active_countries", "CONFIG_SHOW_ACTIVE_COUNTRIES")
	viper.BindEnv("config.cleanup_battles_days_old", "CONFIG_CLEANUP_BATTLES_DAYS_OLD")
	viper.BindEnv("config.cleanup_guests_days_old", "CONFIG_CLEANUP_GUESTS_DAYS_OLD")
	viper.BindEnv("config.cleanup_retros_days_old", "CONFIG_CLEANUP_RETROS_DAYS_OLD")
	viper.BindEnv("config.cleanup_storyboards_days_old", "CONFIG_CLEANUP_STORYBOARDS_DAYS_OLD")
	viper.BindEnv("config.organizations_enabled", "CONFIG_ORGANIZATIONS_ENABLED")

	viper.BindEnv("feature.poker", "FEATURE_POKER")
	viper.BindEnv("feature.retro", "FEATURE_RETRO")
	viper.BindEnv("feature.storyboard", "FEATURE_STORYBOARD")

	viper.BindEnv("auth.method", "AUTH_METHOD")
	viper.BindEnv("auth.ldap.url", "AUTH_LDAP_URL")
	viper.BindEnv("auth.ldap.use_tls", "AUTH_LDAP_USE_TLS")
	viper.BindEnv("auth.ldap.bindname", "AUTH_LDAP_BINDNAME")
	viper.BindEnv("auth.ldap.bindpass", "AUTH_LDAP_BINDPASS")
	viper.BindEnv("auth.ldap.basedn", "AUTH_LDAP_BASEDN")
	viper.BindEnv("auth.ldap.filter", "AUTH_LDAP_FILTER")
	viper.BindEnv("auth.ldap.mail_attr", "AUTH_LDAP_MAIL_ATTR")
	viper.BindEnv("auth.ldap.cn_attr", "AUTH_LDAP_CN_ATTR")

	err := viper.ReadInConfig()
	if err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); !ok {
			logger.Fatal(err.Error())
		}
	}
}
